### PR TITLE
E2E composability testing in Pytorch core repo

### PR DESCRIPTION
**Pull request by mori360**
Add 3D tests(pp, tp, dp) at Pytorch based on 3D tests at Torchtitan
1. Initialize the 3D test model here.
- pp (with 5 different pp schedules)
- tp (async = True/False)
- dp (fsdp, ddp)
- compile
2. The current model could run with the config combinations above, but we focus on 3d here. Open to add other tests
3. All the APIs are based on pytorch, could run independent from Torchtitan

Todo:
1. enable checkpoint
2. add pp loss, weight.grad test
4. enable float8 when it's available

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 1]
```diff
@@ -0,0 +1,757 @@
...
+        model_args = ModelArgs(dropout_p=0.0)
+        model = Transformer(model_args)
+        world_mesh = parallel_dims.build_mesh(device_type="cuda")
+
+        if parallel_dims.float8:
+            scale_for_fsdp = False
+            if parallel_dims.dp_enabled and parallel_dims.dp_type == "fsdp":
+                scale_for_fsdp = True
+            # a no-op hander if float8 is not enabled
+            float8_handler = Float8Handler(parallel_dims, scale_for_fsdp)
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:293]: is float8 path tested in CI? it requires dependency on torchao

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+                ],
+                "tp_degree": [
+                    2,
+                ],
+                "pp_degree": [
+                    2,
+                ],
+                "float8": [True],
+                "async_tp": [True],
+                "schedule_class": ["1f1b"], #["1f1b","gpipe","interleaved_1f1b","flexible_interleaved_1f1b",]
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:246]: how to we decide which schedule to run?

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+                "dp_degree": [
+                    1,
+                ],
+                "tp_degree": [
+                    2,
+                ],
+                "pp_degree": [
+                    2,
+                ],
+                "float8": [True],
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:244]: float8 = [True, False]

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+                "async_tp": [True],
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:245]: async_tp = [True, False]

File: test/distributed/_composable/test_composability/test_3d_composability.py (hunk 1 above)
Comment (mori360) [reply test/distributed/_composable/test_composability/test_3d_composability.py:293]: It's tested right now, with torchao imported when `float8=True`

File: test/distributed/_composable/test_composability/test_3d_composability.py (hunk 2 above)
Comment (mori360) [reply test/distributed/_composable/test_composability/test_3d_composability.py:246]: The current `run_subtests` is not ideal, would collect feedback and switch to other methods.

File: test/distributed/_composable/test_composability/test_3d_composability.py (hunk 1 above)
Comment (weifengpy) [reply test/distributed/_composable/test_composability/test_3d_composability.py:293]: I mean pytorch CI needs torchao as depdency. I believe local devgpu works because you installed torchao by yourself

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        self.logger.addHandler(ch)
+
+        os.environ["KINETO_LOG_LEVEL"] = "5"
+
+    @skip_if_lt_x_gpu(4)
+    def test_3d_composability(self):
+        self.run_subtests(
+            {
+                "dp_degree": [
+                    1,
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:236]: dp_degree = 1 means disable FSDP?

Comment (mori360) [reply test/distributed/_composable/test_composability/test_3d_composability.py:236]: later would change the subset structure to cover more tests
Currently could run with dp, tp, pp degree in [1,2,4]

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+            )
+            for m in model_parts:
+                self._parallelize_llama(m, world_mesh, parallel_dims)
+                m.to_empty(device="cuda")
+                m.train()
+        else:
+            self._parallelize_llama(model, world_mesh, parallel_dims)
+            model.train()
+            model_parts = [model]
+
```
Comment (weifengpy) [inline test/distributed/_composable/test_composability/test_3d_composability.py:321]: is the next step to add forward/backward and optimizer step? seems we are testing model init right now

File: test/distributed/_composable/test_composability/test_3d_composability.py (hunk 5 above)
Comment (weifengpy) [reply test/distributed/_composable/test_composability/test_3d_composability.py:236]: it's just I remember dp=1 means no dp. so for model init, wnat to make sure fully_shard is indeed called in current setting
